React test submission

dangovan opened the pull request:
Seems to look and work right. Covered a few gotchas but blanket solutions would have been better. Tests are very rudimentary but they pass.

culshaw commented:
Thank you for this!

I'll review and leave some comments.

I have a couple of questions:
- What do you use to run the tests you wrote? `npm run X`
- What version of node did you install the dependencies with? Did you use yarn?

dangovan commented:
Just `npm test`
I'm running Node v11.1.0,  just used `npm install` and `npm start`
I updated React scripts to 2.0 so I could use css modules and scss directly,  and added some enzyme packages as it came recommended for testing

culshaw commented on src/App.js:
```diff
@@ -1,22 +1,111 @@
+
+      let characters = [...this.state.characters];
+      let allCharacters = this.state.cache;
+      //Must be a better way of keeping a backup of the full list of characters than running twice the operations ever time?
```
It shouldn't matter how the original list is ordered if you are changing the rendered output based on a filter changing. If there are 2 filters, you react to the filter change e.g. `sorterChangedHandler ` and call a helper abstracted function which processes all your current filter states.


```
this.state = {
	characters: [list of characters which never gets updated]
	filteredCharacters: [] // this will be constantly updated
};
...

sorterChangedHandler: function(event) {
	const filters = {...this.state.filters, sort: event.target.value };

	this.setState({
		filters 
	});

	this.filterMyList(this.state.characters, filters);
}

filterMyList: function(characters, filters) {
	// Sort
 	characters.sort((a, b) => a.significanceIndex - b.significanceIndex);
    // Filter
	this.alphaSorter(characters)

	this.setState({
		filteredCharacters: characters
	});
}

```

culshaw commented on src/components/CharacterEntry/CharacterEntry.js:
```diff
@@ -0,0 +1,14 @@
+
+const characterEntry = (props) => (
+  <li className={classes.CharacterEntry} >
+    <img src={/characters/ + props.avatar} alt={props.name} />
```
This is interesting, `{/characters/ + props.avatar}` I would expect this to evaluate as a RegEx. Which would be really strange for a src prop. 

I suspect this should be: `<img src={"/characters/" + props.avatar} alt={props.name} />`

culshaw commented:
`npm test` won't work for me with node 11.10 :(

culshaw commented on src/components/FilterSelector/FilterSelector.test.js:
```diff
@@ -0,0 +1,13 @@
+
+describe('<NavigationItems />', () => {
+  it("renders without crashing", () => {
+      const wrapper = shallow(<FilterSelector options={[test, test]} />);
```
I can't get it to run but I'm pretty sure this would fail too: `[test, test]`

dangovan commented on src/App.js:
```diff
@@ -1,22 +1,111 @@
+      characters,
+      cache: {},
+      touched: false,
+      categories: [test, test],
```
Removed the undefined 'test's, odd that it didn't break mine

dangovan commented:
> `npm test` won't work for me with node 11.1.0 :(

npm run-script test does the same for me, i must have aliases
